src/validation.cpp

```diff
@@ -5050,19 +5050,25 @@ void ChainstateManager::CheckBlockIndex()
         return;
     }
 
-    // Build forward-pointing map of the entire block tree.
+    // Build forward-pointing data structure for the entire block tree.
+    // For performance reasons, indexes of the best header chain are stored in a vector (within CChain).
+    // All remaining blocks are stored in a multimap.
+    CChain best_hdr_chain;
+    if (m_best_header) best_hdr_chain.SetTip(*m_best_header);
```

Comment:
In 4b05c654c1:

Using `m_best_header` shouldn't be working (at least not as you expect) for `invalidateblock`.

During `InvalidateBlock()`, we first disconnect all blocks, then call `CheckBlockIndex` without updating `m_best_header` (this is done later on the flow, when we call `InvalidChainFound()`). So, by the time `CheckBlockIndex` is called, `best_hdr_chain` is loaded with a chain that is not the active one.

Any reason not to use `ActiveTip()` instead of `m_best_header` here?

Reply:
> Any reason not to use `ActiveTip()` instead of `m_best_header` here?

I think the reason is just that we want `best_hdr_chain` to be the longest chain with the most forks for the optimization to work well. We don't actually care which chain it is because CheckBlockIndex will traverse all the blocks in any case. If a different chain is chosen it just means the blocks will be traversed in a different order, less or more efficiently.

Reply:
ok, yeah. Then it is not an issue but, isn't error-prone to call `best_hdr_chain` something that might not be the best chain then? E.g. it seems that if we ever add a check only for the active chain blocks here, the `best_hdr_chain.Contains()` usage temptation will be there.

Reply:
> ok, yeah. Then it is not an issue but, isn't error-prone to call `best_hdr_chain` something that might not be the best chain then? E.g. it seems that if we ever add a check only for the active chain blocks here, the `best_hdr_chain.Contains()` usage temptation will be there.

I don't love the name `best_hdr_chain` but don't think it is too bad. I agree it could potentially be confused with the active chain in new code, and that would be a bug. In my original suggestion https://github.com/bitcoin/bitcoin/pull/28339#discussion_r1522152244, I called it `best_headers` so at least the name was consistent with `m_best_header`. I don't really have other ideas though.

Reply:
> Any reason not to use ActiveTip() instead of m_best_header here

My main reason is that `CheckBlockIndex()` being independent from the active chainstate makes the performance improvement applicable to situations where we have all the headers but haven't validated much of the chain: See  [this old answer](https://github.com/bitcoin/bitcoin/pull/28339#discussion_r1307561848). IBD and reindex seem much more relevant than extreme `invalidateblock` scenarios.

Reply:
> I agree it could potentially be confused with the active chain in new code, and that would be a bug.

I added a comment in `CheckBlockIndex()` describing why these can differ.